crates/bevy_ecs/src/world/entity_ref.rs

```diff
@@ -2107,6 +2107,248 @@ impl<'w> EntityWorldMut<'w> {
+    pub fn clone_components_with_requires<B: Bundle>(&mut self, target: Entity) -> &mut Self {
+        self.assert_not_despawned();
+
+        let storages = &mut self.world.storages;
+        let components = &mut self.world.components;
+        let bundles = &mut self.world.bundles;
+
+        let bundle_id = bundles.register_contributed_bundle_info::<B>(components, storages);
+        // SAFETY: the `BundleInfo` for this `BundleId` is initialized above
+        let bundle_info = unsafe { bundles.get_unchecked(bundle_id) };
+        let component_ids = bundle_info.contributed_components().to_owned();
+
+        let mut builder = EntityCloneBuilder::new(self.world);
+        builder.deny_all().allow_by_ids(component_ids);
+        builder.clone_entity(self.entity, target);
+
+        self.world.flush();
+        self.update_location();
+        self
+    }
+
+    /// Clones the given components of this entity and inserts them into another entity.
+    ///
+    /// Components can only be cloned if they implement
+    /// [`Clone`] or [`Reflect`](bevy_reflect::Reflect).
+    ///
+    /// You should prefer to use the typed API [`EntityWorldMut::clone_components`] where possible.
+    ///
+    /// # Panics
+    ///
+    /// - If this entity has been despawned while this `EntityWorldMut` is still alive.
+    /// - If the target entity does not exist.
+    pub fn clone_components_by_id(
+        &mut self,
+        target: Entity,
+        ids: impl IntoIterator<Item = ComponentId>,
+    ) -> &mut Self {
+        self.assert_not_despawned();
+
+        let mut builder = EntityCloneBuilder::new(self.world);
+        builder.deny_all().allow_by_ids(ids);
+        builder.clone_entity(self.entity, target);
+
+        self.world.flush();
+        self.update_location();
+        self
+    }
+
+    /// Clones the specified components of this entity and inserts them into another entity,
+    /// then removes the components from this entity.
+    ///
+    /// Components can only be cloned if they implement
+    /// [`Clone`] or [`Reflect`](bevy_reflect::Reflect).
+    ///
+    /// # Panics
+    ///
+    /// - If this entity has been despawned while this `EntityWorldMut` is still alive.
+    /// - If the target entity does not exist.
+    pub fn move_components<B: Bundle>(&mut self, target: Entity) -> &mut Self {
+        self.assert_not_despawned();
+
+        let mut builder = EntityCloneBuilder::new(self.world);
+        builder.deny_all().allow::<B>();
+        builder.clone_entity(self.entity, target);
+
+        let storages = &mut self.world.storages;
+        let components = &mut self.world.components;
+        let bundles = &mut self.world.bundles;
+        let bundle_id = bundles.register_info::<B>(components, storages);
+
+        // SAFETY: the `BundleInfo` for this `BundleId` is initialized above
+        unsafe { self.remove_bundle(bundle_id) };
+
+        self.world.flush();
+        self.update_location();
+        self
+    }
```

Comment:
Moving components can also be a configuration option for `EntityCloneBuilder` that would call `remove_bundle` as a final step in `clone_entity`. Theoretically, it would also allow us to add an optimization in the future that would allow `EntityCloner` to move components that don't implement `Clone` or `Reflect`.